source/simulator/assembly.cc

```diff
@@ -1551,6 +1636,906 @@ namespace aspect
     computing_timer.exit_section();
   }
 
+  template <int dim>
+  void Simulator<dim>::
+  local_assemble_advection_face_terms(const AdvectionField     &advection_field,
+                                      const typename DoFHandler<dim>::active_cell_iterator &cell,
+                                      internal::Assembly::Scratch::AdvectionSystem<dim> &scratch,
+                                      internal::Assembly::CopyData::AdvectionSystem<dim> &data)
+  {
+    const bool use_bdf2_scheme = (timestep_number > 1);
+
+    const unsigned int n_q_points    = scratch.face_finite_element_values.n_quadrature_points;
+
+    // also have the number of dofs that correspond just to the element for
+    // the system we are currently trying to assemble
+    const unsigned int advection_dofs_per_cell = data.local_dof_indices.size();
+
+    Assert (advection_field.is_discontinuous(introspection), ExcMessage("Face terms only need assembling in the case of discontinuous temperature or compositional discretization."));
+
+    Assert (advection_dofs_per_cell < scratch.face_finite_element_values.get_fe().dofs_per_cell, ExcInternalError());
+    Assert (scratch.face_grad_phi_field.size() == advection_dofs_per_cell, ExcInternalError());
+    Assert (scratch.face_phi_field.size() == advection_dofs_per_cell, ExcInternalError());
+
+    const unsigned int solution_component
+      = (advection_field.is_temperature()
+         ?
+         introspection.component_indices.temperature
+         :
+         introspection.component_indices.compositional_fields[advection_field.compositional_variable]
+        );
+
+    const FEValuesExtractors::Scalar solution_field
+      = (advection_field.is_temperature()
+         ?
+         introspection.extractors.temperature
+         :
+         introspection.extractors.compositional_fields[advection_field.compositional_variable]
+        );
+
+    //loop over all possible subfaces of the cell, and reset their matrices.
+    for (unsigned int f = 0; f < GeometryInfo<dim>::max_children_per_face * GeometryInfo<dim>::faces_per_cell; ++f)
+      {
+        data.local_matrices_ext_int[f] = 0;
+        data.local_matrices_int_ext[f] = 0;
+        data.local_matrices_ext_ext[f] = 0;
+      }
+
+    for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
+      {
+
+        typename DoFHandler<dim>::face_iterator face = cell->face (f);
+        if (face->at_boundary())
+          {
+            scratch.face_finite_element_values.reinit (cell, f);
+
+            scratch.face_finite_element_values[introspection.extractors.velocities].get_function_values(current_linearization_point,
```

Comment:
can't you move all of these initializations into the `if` statement below that checks that it is indeed a Dirichlet-type boundary?